### ISSUE TITLE
Custom Trainning 2900 hours

### ISSUE THREAD

**Issue by rebotnix** [Rudrabha/Wav2Lip#75]
We would like to train a HD dataset where we create 2900 hours of a custom training video dataset. I think the training will take several weeks, even with 4 x v100 cards. 

We prepare the dataset with an face detection and lip-detection model, so that the videos only have visible face and mouth movements. Is it correct that each single frame/jpg is a full frame or does it make a change when we cut out the mouth region?

Here is a example of the fullframe. Is it correct?
Each video has this start position, end, as well as several sides, face views so i hope i will get better results then the pre-trained mode.

![Bildschirmfoto 2020-09-22 um 22 27 48](https://user-images.githubusercontent.com/566761/93934601-b294cb80-fd23-11ea-9b10-a3e10d4405ab.jpg)

I take a look at the hparams.img_size and it is a very small size that i would like to change at first to 320px, later 720p and maybe 1080p.

I will share the model as soon it is ready trained. If anyone has a tip for starting, please let me know.

**Comment by prajwalkr** [comment]
> or does it make a change when we cut out the mouth region?

You can detect faces and crop them out and train on them. All recent works have done that. Most importantly, the speech only corresponds to the lip region. You are asking the model to unnecessarily generate the rest of the background if you do not cut out the face and feed it. Your training will also be much faster if you crop out faces.

**Comment by rebotnix** [comment]
Very good point, I can run the SFD model for face extraction and save the image. If you like, please let the issue open so that i can share my result for all other maintainers here.

**Comment by prajwalkr** [comment]
Great! You can check here for some tips: https://github.com/Rudrabha/Wav2Lip#training-on-datasets-other-than-lrs2. I will leave the issue open :+1:

**Comment by quangbang1989** [comment]
@rebotnix how do cantact you i want to ask you something please

**Comment by wangrui5781** [comment]
I trained the model in more then 4 dataset and solved some problems in the experiment.
1. The output image exactly like the random frame. It seems like perceptual gradient more stronger then sync gradient, so I trained the model only contrain restrained by L1 loss. And once the L1 loss less then 0.3 and eval sync loss less then 2, I add a slight sync weight 0.002. About 15K step later, L1 loss archive 0.25 and sync loss archive 0.5 . Then perceptual gan joined in the trainning and become more stable.
2. Gradient explosion of perceptual gan. The observation is perceptual loss archive 0 and Fake more then 27. I make a sharp to the perceptual discriminator, change the real target to 0.9 and the fake target to 0.1 and resolve it. The code is : 

disc_real_loss = F.binary_cross_entropy(pred, torch.ones((len(pred), 1)).to(device) * 0.9)
disc_fake_loss = F.binary_cross_entropy(pred, torch.ones((len(pred), 1)).to(device) * 0.1)

in the train.py and 

        false_pred_loss = F.binary_cross_entropy(self.binary_pred(false_feats).view(len(false_feats), -1), 
                                        torch.ones((len(false_feats), 1)).cuda() * 0.9)

in wav2lip.py.
Sorry for my poor English and I hope these experiments can help you.

**Comment by rebotnix** [comment]
I start creating sharing an HD dataset where we publish several youtube AV datasets videos and a script that handle and prepare all files. 

This script downloads all youtube video from the CSV (will upload it today) and cuts it with to the right speech segments FFmpeg. I will share the face cut part also very soon. 

Patch
https://github.com/Rudrabha/Wav2Lip/commit/d9ef10a9977e5c2f7ae97464e358c81678176459

**Comment by quangbang1989** [comment]
you are welcome, your works are perfect, i am staying here and waiting your
sharing.
Thank you so much.

**Comment by rebotnix** [comment]
In my fork, there are some experimental code that you can use. When it makes sense prajwalkr will for sure merge I think.

Test it here:
https://github.com/rebotnix/Wav2Lip/tree/master/utils

We are implementing the face extraction very soon in the grap.py script.

**Comment by JonathanSum** [comment]
I am just asking. If I want to train it on another one-hour long video(or we call it dataset here), I need to crop out the faces? if so, how do I train it and use it in this wave to lip?

**Comment by rebotnix** [comment]
> I am just asking. If I want to train it on another one-hour long video(or we call it dataset here), I need to crop out the faces? if so, how do I train it and use it in this wave to lip? And that is for the discriminator to determine the image is fake or not?

You can use the preprocessing, preprocess.py that crops all for you including extracting the audio from the video.

**Comment by JonathanSum** [comment]
It looks like this repo says if the lips in the video is not aligned, I have to retrain it.

**Comment by Kemito** [comment]
@rebotnix Apologies for re-opening here but your FORK seems to be set to private. Is there any progress update? Did you manage to train it to HD? Mentioned that you might share the result of the model. I would be interested to see the result. Please let me know if it is possible.

**Comment by MrWig** [comment]
> @rebotnix Apologies for re-opening here but your FORK seems to be set to private. Is there any progress update? Did you manage to train it to HD? Mentioned that you might share the result of the model. I would be interested to see the result. Please let me know if it is possible.

Hopping on this bandwagon as well as I have been experimenting with implementing this repo and would benefit from trying out a few forks or training on some other sets.

**Comment by mushanshanshan** [comment]
Hi @rebotnix , I'm also trying to re-training an HD model on avspeech dataset, it seems that your fork was set as private, would you mind sharing some experience or code?
